timemachine/cpp/src/kernels/k_electrostatics.cuh

```diff
@@ -1,104 +1,78 @@
 #pragma once
-
 #include "kernel_utils.cuh"
 
-template<typename RealType>
+template<typename RealType, size_t NDIMS>
 void __global__ k_electrostatics(
     const int num_atoms,    // n
-    const RealType *coords, // [n, 3]
+    const RealType *coords, // [n, D]
     const RealType *params, // [p,]
     const RealType *scale_matrix, // [n, n]
     const int *param_idxs,  // [n, 1] charge
     RealType *E,             // [,] or null
-    RealType *dE_dx,         // [n,3] or null
-    RealType *d2E_dx2,       // [C, n, 3, n, 3] or null, hessian
+    RealType *dE_dx,         // [n,D] or null
+    RealType *d2E_dx2,       // [C, n, D, n, D] or null, hessian
     // parameters used for computing derivatives
     const int num_dp,        // dp, number of parameters we're differentiating w.r.t. 
     const int *param_gather_idxs, // [p,] if -1, then we discard
     RealType *dE_dp,         // [C, dp,] or null
-    RealType *d2E_dxdp       // [C, dp, n, 3] or null
+    RealType *d2E_dxdp       // [C, dp, n, D] or null
 ) {
 
     const auto conf_idx = blockIdx.z;
     const int N = num_atoms;
     const int DP = num_dp;
 
     auto i_idx = blockDim.x*blockIdx.x + threadIdx.x;
-    
-    RealType x0, y0, z0, q0;
+    RealType q0;
+    RealType X0[NDIMS] = {0};
     int q0_g_idx;
 
     if(i_idx >= N) {
-        x0 = 0.0;
-        y0 = 0.0;
-        z0 = 0.0;
         q0 = 0.0;
         q0_g_idx = 0;
     } else {
-        x0 = coords[conf_idx*N*3+i_idx*3+0];
-        y0 = coords[conf_idx*N*3+i_idx*3+1];
-        z0 = coords[conf_idx*N*3+i_idx*3+2];
+        #pragma unroll
```

Comment:
Is NDIMS really const here?  I’m unsure if the compiler can tell.

Reply:
it's templated to instances of ndims 3 and 4